test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs

```diff
@@ -235,7 +235,8 @@ public void LineString_Intersection_with_Polygon_return_null1()
             Debug.WriteLine(res.AsText());
         }
 
-        [Test(Description = "GitHub Issue #125")]
+        [Test(Description = "GitHub Issue #124")]
+        [Category("FailureCase")]
```

Comment:
We should still figure out why this test fails... is it a bug upstream?

Reply:
This is because the polygon is actually a figure-8, and buffer(0) often doesn't handle those correctly.

This test probably isn't relevant any more, because `GeometryPrecisionReducer` now uses Snap-Rounding, which does not use buffer(0) to creating valid topology.

Reply:
@dr-jts the exception is actually thrown in `GeometryPrecisionReducer` itself, specifically [here](https://github.com/NetTopologySuite/NetTopologySuite/blob/bead06f69f2cfab88dde7df96d2b4b351abeeaf4/src/NetTopologySuite/Operation/OverlayNG/PolygonBuilder.cs#L190).  I haven't tried this in JTS to determine whether this is a problem with the algorithm or with our port.

⌛ if I get the time to get around to it, I'll try it out and report it in JTS if it fails.

Reply:
Ah, I understand now.  JTS throws the same error.  And in fact it throws for any precision value.  The problem is that the input geometry is invalid.  SInce `GeometryPrecisionReducer` now uses OverlayNG, it is more sensitive to invalidity.  This might be considered a regression, I suppose - but it still seems like a reasonable & consistent contract that input needs to be valid.  

The usual heuristic to fix bad geometry is to use `buffer(0)`, but in this case as noted that fails, producing a `POLYGON EMPTY`.  It is only a heuristic, and never guaranteed to work.  At some point hopefully JTS will acquire a `MakeValid` function.

Reply:
The cause of the invalidity is a "bowtie" self-intersection at the right end of the polygon.  The points forming it are only on the order of 1e-15 apart!

![image](https://user-images.githubusercontent.com/3529053/94471973-e71ee200-017e-11eb-8da9-001850cec774.png)

It is interesting that this fails, since you would expect these points to snap together.  This might actual reveal a flaw in the OverlayNG algorithm, in that it uses the original polygon to check for containment, which leaves it susceptible to input invalidity.  I should probably revisit this design.

Reply:
The reason this case fails in `GeometryPrecisionReducer` is because of the invalid input.  Because of the bowtie polygon, the orientation of the ring is ambiguous.  It is evaluated in such a way that the large portion of the ring has orientation indicating that it is a hole.  The smaller "shell" part collapses to a single point, leaving just a hole ring.  This then of course cannot be assigned to a shell.

The throwing of a `TopologyException` is actually a good thing, since it indicates there is something wrong with the input.   It's a bit more worrying that this polygon can actually produce output when used in say intersections (although the output is invalid).  At the moment I don't have an idea of how to improve this.